### PR TITLE
JUnit XML for Unit Tests and Bug Fix

### PR DESCRIPTION
**Pull request by chxchx**
The bug was that when a Prow job fails, no xml results are uploaded because the test script exit on nonzero status immediately and the xml file was move to artifacts / upload directory as the last step. By creating the files in `_artifacts` dir directly, it is always uploaded by Prow. 

Also git ignores the junit results so developers running tests locally will not pollute his/her workspace. Passing junit xml file names through environment variables provides more flexibility and prevents overwriting results if executing multiple commands that all generate junit xmls.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -380,7 +380,10 @@ istioctl-install:
 .PHONY: test localTestEnv test-bins
 
 # Run coverage tests
-test: pilot-test mixer-test security-test broker-test galley-test common-test
+test:
+	set -o pipefail; \
+	$(MAKE) pilot-test mixer-test security-test broker-test galley-test common-test \
+	|& tee >(go-junit-report > junit.xml)
```
Comment (mattdelco) [inline Makefile]: I can't help but wonder if dumping to junit.xml will eventually be a problem, either because it'll by default dirty the source tree (e.g., it could be stored under "${ISTIO_OUT}/junit.xml" ) or because there will now be multiple targets using the same name (which would mostly matter for parallel execution [or possibly serial if someone forgets to move/rename the file between running different tests).  Anyway, just food for thought (vs an expected change).

Comment (chxchx) [reply Makefile]: Those are my concerns too. I started the PR as an experiment and as a follow up I let the caller supply a destination file in place of junit.xml to address the concerns. Please take another look.

File: .gitignore [hunk 2]
```diff
@@ -50,3 +50,5 @@ install/kubernetes/istio.yaml
 samples/bookinfo/consul/bookinfo.sidecars.yaml
 samples/bookinfo/eureka/bookinfo.sidecars.yaml
 istio.VERSION
+# test results in junit format
```
Comment (mattdelco) [inline .gitignore]: can revert the changes to this file

File: Makefile [hunk 3]
```diff
@@ -380,7 +380,11 @@ istioctl-install:
 .PHONY: test localTestEnv test-bins
 
 # Run coverage tests
-test: pilot-test mixer-test security-test broker-test galley-test common-test
+JUNIT_UNIT_TEST_XML ?= junit_unit_tests.xml
```
Comment (mattdelco) [inline Makefile]: JUNIT_UNIT_TEST_XML ?= $(ISTIO_OUT)/junit_unit_tests.xml

File: Makefile [hunk 4]
```diff
@@ -380,7 +380,11 @@ istioctl-install:
...
+test:
+	set -o pipefail; \
```
Comment (mattdelco) [inline Makefile]: mkdir -p $(notdir $(JUNIT_UNIT_TEST_XML))

File: prow/istio-unit-tests.sh [hunk 5]
```diff
@@ -27,9 +27,11 @@ set -e
...
+cd ${ROOT}
+
 echo 'Build'
-(cd ${ROOT}; make build)
+make build
 
 # Unit tests are run against a local apiserver and etcd.
 # Integration/e2e tests in the other scripts are run against GKE or real clusters.
-(cd ${ROOT}; make localTestEnv test)
+JUNIT_UNIT_TEST_XML="${ARTIFACTS_DIR}/junit_unit_tests.xml" make localTestEnv test
```
Comment (mattdelco) [inline prow/istio-unit-tests.sh]: Could add "build" before "localTestEnv" and get rid of the separate "make build" invocation done a few lines above.

File: prow/lib.sh [hunk 6]
```diff
@@ -39,6 +39,7 @@ function setup_and_export_git_sha() {
 
     # Set artifact dir based on checkout
     export ARTIFACTS_DIR="${GOPATH}/src/istio.io/istio/_artifacts"
+    mkdir -p "${ARTIFACTS_DIR}"
```
Comment (mattdelco) [inline prow/lib.sh]: This change can probably go away if the Makefile does the mkdir.

File: tests/istio.mk [hunk 7]
```diff
@@ -84,8 +84,11 @@ e2e_bookinfo: istioctl generate_yaml
 e2e_upgrade: istioctl generate_yaml
 	go test -v -timeout 20m ./tests/e2e/tests/upgrade -args ${E2E_ARGS} ${EXTRA_E2E_ARGS}
 
+JUNIT_E2E_XML ?= junit.xml
```
Comment (mattdelco) [inline tests/istio.mk]: JUNIT_E2E_XML ?= $(ISTIO_OUT)/junit.xml

File: tests/istio.mk [hunk 8]
```diff
@@ -84,8 +84,11 @@ e2e_bookinfo: istioctl generate_yaml
...
 e2e_all:
-	set -o pipefail; $(MAKE) e2e_simple e2e_mixer e2e_bookinfo |& tee >(go-junit-report > junit.xml)
```
Comment (mattdelco) [inline tests/istio.mk]: mkdir -p $(dir $(JUNIT_E2E_XML))

File: Makefile (hunk 4 above)
Comment (chxchx) [reply Makefile]: I think you meant to use dir instead of notdir ?

Comment (mattdelco) [reply Makefile]: oh, yeah, sorry.

**Comment by chxchx** [comment]
Agree with proposal from @mattdelco to use $ISTIO_OUT as default location and call mkdir from make recipes.

File: Makefile [hunk 9]
```diff
@@ -387,7 +387,12 @@ istioctl-install:
 .PHONY: test localTestEnv test-bins
 
 # Run coverage tests
-test: pilot-test mixer-test security-test broker-test galley-test common-test
+JUNIT_UNIT_TEST_XML ?= $(ISTIO_OUT)/junit_unit_tests.xml
+test: | $(ISTIO_OUT)
```
Comment (mattdelco) [inline Makefile]: You don't need to (and probably shouldn't) add the "| $(ISTIO_OUT)" since you've added the mkdir.  The "| $(ISTIO_OUT)" basically just causes make to create the directory if it doesn't already exist.

File: tests/istio.mk [hunk 10]
```diff
@@ -84,8 +84,12 @@ e2e_bookinfo: istioctl generate_yaml
 e2e_upgrade: istioctl generate_yaml
 	go test -v -timeout 20m ./tests/e2e/tests/upgrade -args ${E2E_ARGS} ${EXTRA_E2E_ARGS}
 
-e2e_all:
-	set -o pipefail; $(MAKE) e2e_simple e2e_mixer e2e_bookinfo |& tee >(go-junit-report > junit.xml)
+JUNIT_E2E_XML ?= $(ISTIO_OUT)/junit.xml
+e2e_all: | $(ISTIO_OUT)
```
Comment (mattdelco) [inline tests/istio.mk]: Same here about "| $(ISTIO_OUT)" being superfluous since the "mkdir" was added.

**Comment by mattdelco** [comment]
/lgtm
but my PR says expect the tests to fail for unrelated reasons (something wrong with updating vendor/)

**Comment by ldemailly** [comment]
/test e2e-bookInfo
should pass now

**Comment by ldemailly** [comment]
> but my PR says expect the tests to fail for unrelated reasons (something wrong with updating vendor/)

that should be fixed now, sorry about that

File: tests/istio.mk [hunk 11]
```diff
@@ -84,8 +84,12 @@ e2e_bookinfo: istioctl generate_yaml
 e2e_upgrade: istioctl generate_yaml
 	go test -v -timeout 20m ./tests/e2e/tests/upgrade -args ${E2E_ARGS} ${EXTRA_E2E_ARGS}
 
+JUNIT_E2E_XML ?= $(ISTIO_OUT)/junit.xml
```
Comment (costinm) [inline tests/istio.mk]: Well, can't be just junit.xml - make it junit_e2e_all.xml or something.

And I think it'll be good to have each separate step generate an xml too - since we want
to run them in parallel ( at least in some CIs )

File: Makefile [hunk 12]
```diff
@@ -381,7 +381,12 @@ istioctl-install:
 .PHONY: test localTestEnv test-bins
 
 # Run coverage tests
-test: pilot-test mixer-test security-test broker-test galley-test common-test
+JUNIT_UNIT_TEST_XML ?= $(ISTIO_OUT)/junit_unit_tests.xml
```
Comment (costinm) [inline Makefile]: Please also update .circleci to use this target if possible

Comment (chxchx) [reply Makefile]: It is the same `test` target that circle ci has been using. By updating the target implementation, it always generates the junit results regardless of caller.

File: tests/istio.mk (hunk 11 above)
Comment (chxchx) [reply tests/istio.mk]: Done

**Comment by chxchx** [comment]
PTAL

It is very cool that Prow UI also changes given the junit results. Once this PR is in, testgrid will render per test case pass/fail info from each run.

File: Makefile [hunk 13]
```diff
@@ -381,7 +381,12 @@ istioctl-install:
...
+test:
```
Comment (mattdelco) [inline Makefile]: "test" is part of the default rule, so I'm afraid that this will break for devs that don't happen to already have "go-junit-report" installed.  It seems like there might need to be a rule/dependency install this if necessary, or instead not set a default and this make it opt-in behavior (by setting JUNIT_UNIT_TEST_XML), where the tee part can probably be made conditional using something like:
<pre>$(if $(JUNIT_UNIT_TEST_XML),|& tee >(go-junit-report > $(JUNIT_UNIT_TEST_XML)) )</pre>
or I suppose another option is to have a targets for the opt-in of junit behavior.

Comment (chxchx) [reply Makefile]: I added a new target that check if go-junit-report is installed and install it if not.

**Comment by mattdelco** [comment]
I think the approach for go get would only work if someone has their PATH set up.  A scheme I've been using elsewhere for other tools is to first have a variable that says where to find the parser, e.g.:
<pre>JUNIT_REPORT:=$(shell which go-junit-report 2> /dev/null || echo "${ISTIO_BIN}/go-junit-report")</pre>
and then have a rule that tells Make how to download the tool if necessary:
<pre>
${ISTIO_BIN}/go-junit-report:
	@echo "go-junit-report not found. Installing it now..."
	unset GOOS && CGO_ENABLED=1 go get -u github.com/jstemmer/go-junit-report
</pre>
and then have an order-only dependency on $(JUNIT_REPORT) (so that Make doesn't throw a fit about not knowing how to build the instance that might already exist elsewhere on the system) and then use "$(JUNIT_REPORT)" instead of "go-junit-report" in the build recipe.

**Comment by mattdelco** [comment]
/lgtm
A possible refinement is to also use "&& unset GOARCH" with the
<pre>unset GOOS && CGO_ENABLED=1 go get -u github.com/jstemmer/go-junit-report</pre>
though it's not a big deal yet.
